react/features/video-layout/subscriber.js

```diff
@@ -4,21 +4,106 @@ import {
 
 /**
  * StateListenerRegistry provides a reliable way of detecting changes to
  * preferred layout state and dispatching additional actions.
  */
 StateListenerRegistry.register(
     /* selector */ state => shouldDisplayTileView(state),
-    /* listener */ (displayTileView, { dispatch }) => {
+    /* listener */ (displayTileView, store) => {
+        const { dispatch } = store;
+
         dispatch(selectParticipant());
 
         if (!displayTileView) {
-            dispatch(setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+            dispatch(
+                setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+
+            if (typeof interfaceConfig === 'object'
+                && interfaceConfig.AUTO_PIN_LATEST_SCREEN_SHARE) {
+                _updateAutoPinnedParticipant(store);
+            }
+        }
+    }
+);
+
+/**
+ * For auto-pin mode, listen for changes to the known media tracks and look
+ * for updates to screen shares.
+ */
+StateListenerRegistry.register(
+    /* selector */ state => state['features/base/tracks'],
+    /* listener */ (tracks, store) => {
+        if (typeof interfaceConfig !== 'object'
+            && !interfaceConfig.AUTO_PIN_LATEST_SCREEN_SHARE) {
+            return;
+        }
+
+        const oldScreenSharesOrder
+            = store.getState()['features/video-layout'].screenShares || [];
+        const knownSharingParticipantIds = tracks.reduce((acc, track) => {
+            if (track.mediaType === 'video' && track.videoType === 'desktop') {
+                acc.push(track.participantId);
+            }
+
+            return acc;
+        }, []);
+
+        // Filter out any participants which are no longer screen sharing
+        // by looping through the known sharing participants and removing any
+        // participant IDs which are no longer sharing.
+        const newScreenSharesOrder = oldScreenSharesOrder.filter(
+            participantId => knownSharingParticipantIds.includes(participantId));
+
+        // Make sure all new sharing participant get added to the end of the
+        // known screen shares.
+        knownSharingParticipantIds.forEach(participantId => {
+            if (!newScreenSharesOrder.includes(participantId)) {
+                newScreenSharesOrder.push(participantId);
+            }
+        });
+
+        if (!equals(oldScreenSharesOrder, newScreenSharesOrder)) {
+            store.dispatch(
+                setParticipantsWithScreenShare(newScreenSharesOrder));
+
+            _updateAutoPinnedParticipant(store);
         }
     }
 );
+
+/**
+ * Private helper to automatically pin the latest screen share stream or unpin
+ * if there are no more screen share streams.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {void}
+ */
+function _updateAutoPinnedParticipant({ dispatch, getState }) {
+    const state = getState();
+    const screenShares = state['features/video-layout'].screenShares;
+
+    if (!screenShares) {
```

Comment:
Do we need this check?

Reply:
I'll double check. I think we do because when the store gets initially populated it could be undefined because persistence will restore part of the state (tileViewEnabled) but won't mix state defaults (screenShares). So restoration happens and other things get fired and drop down to this line.

Reply:
Confirmed that's the downstream behavior: the store becomes initialized without screenShares and then several other actions fire, like appWillMount, and end up triggering the tile view subscriber, which ends up triggering this line without the screenShares array.